### PR TITLE
Add option to pass seed into `estimate_u_using_random_sampling`

### LINKED ISSUE #1155: [FEAT] Make `estimate_u_using_random_sampling` reproducible
### Is your proposal related to a problem?

Models aren't currently fully reproducible due to `estimate_u_using_random_sampling` - the u values can be estimated differently each time.

### Describe the solution you'd like

There should be a seed option to ensure the same data and code can reproduce the same sample from which to estimate u.

### Additional context

I think this might be a fairly critical point of failure for splink, seeing as it is the only means of generating u values (I don't believe u can be estimated using EM anymore? Previous versions estimated m, u and lambda using EM, with an option to fix some initial values if desired). Given that, I expect there will be use cases where a suitable sample size is far from clear. U values based on very small subsamples could be highly variable and their accuracy is currently unknown and untested. A single random sample is insufficient for me to be fully confident in the u values being used.

### PR DESCRIPTION
**Pull request by RossKen**
Closes #1155 

Intended to enhance the reproducibility of models.

### CODE REVIEW

**Comment by RossKen** [comment]
Lint fails on `import itertools` in misc.py, saying it is not used. However, it looks to be used in the `all_letter_combos` function at line 84.

File: tests/test_u_train.py [hunk 1]
```diff
@@ -235,3 +235,32 @@ def test_u_train_multilink():
...
+
+    linker_1 = DuckDBLinker(df, settings)
+    linker_2 = DuckDBLinker(df, settings)
+    linker_3 = DuckDBLinker(df, settings)
+
+    linker_1.estimate_u_using_random_sampling(max_pairs=1e3, seed=1)
+    linker_2.estimate_u_using_random_sampling(max_pairs=1e3, seed=1)
+    linker_3.estimate_u_using_random_sampling(max_pairs=1e3, seed=2)
+
+    print(linker_1._settings_obj._parameter_estimates_as_records)
```
Comment (ThomasHepworth) [inline tests/test_u_train.py:257]: rm

Comment (RossKen) [reply tests/test_u_train.py:257]: Good spot - removed

File: splink/athena/athena_linker.py [hunk 2]
```diff
@@ -344,9 +344,14 @@ def register_table(self, input, table_name, overwrite=False):
...
 
-    def _random_sample_sql(self, proportion, sample_size):
+    def _random_sample_sql(self, proportion, sample_size, seed=None):
         if proportion == 1.0:
             return ""
+        if seed:
+            raise NotImplementedError(
+                "Athena does not support seeds in random ",
+                "samples. Please remove the `seed` parameter.",
+            )
```
Comment (ThomasHepworth) [inline splink/athena/athena_linker.py:354]: Do we want this to fail or warn the user that no seed will be set and that u training will proceed?

Comment (RossKen) [reply splink/athena/athena_linker.py:354]: My first thought is for it to fail. Given we aren't setting the seed as a default and people's tendency to ignore warnings - but happy to change if you think it is better to throw a warning instead

Comment (ThomasHepworth) [reply splink/athena/athena_linker.py:354]: Given that the fail is virtually at the start of the process, I'd say it makes sense to leave it as a `NotImplementedError`

**Comment by RossKen** [comment]
@ADBond FYI from this PR you can see that your latest fix for autoblack has now worked 🎉

One thing is once the "lint with black" commit has gone in the tests aren't triggered on the new commit. I'm pretty sure that happened before, right?

**Review by ThomasHepworth — APPROVED** [review]
Thanks Ross, LGTM

**Comment by ThomasHepworth** [comment]
https://github.com/moj-analytical-services/splink/pull/1161#issuecomment-1504045608

On your second point - unfortunately not. I guess ideally we'd have `lint with black` trigger first and then the remaining tests run once that has completed.

**Comment by ThomasHepworth** [comment]
We could also now add an autolinter as ruff comes with the `--fix` argument.

**Comment by ADBond** [comment]
> One thing is once the "lint with black" commit has gone in the tests aren't triggered on the new commit. I'm pretty sure that happened before, right?
Yeah that's right - any commit that is pushed from a github workflow will not trigger other workflows triggered by push/pull request etc. This is so that you don't end up in a situation where you have an infinite loop of workflows. I think we can work around that by creating and using a personal-access-token to push, instead of the default `GITHUB_TOKEN`.

**Pull request merged**